scripts/dts/edtsdatabase.py

```diff
@@ -0,0 +1,193 @@
+import json
+
+##
+# @brief Extended DTS database
+#
+# Database schema:
+#
+# _edts dict(
+#   'devices':  dict(device-id :  device-struct),
+#   'compatibles':  dict(compatible : sorted list(device-id)),
+#   'device-types': dict(device-type : sorted list(device-id)),
+#   ...
+# )
+#
+# device-struct: dict(
+#   'device-id' : device-id,
+#   'device-type' : list(device-type) or device-type,
+#   'compatible' : list(compatible) or compatible,
+#   'label' : label,
+#   property-name : property-value ...
+# )
+#
+# Database types:
+#
+# device-id: <compatible>_<base address in hex>_<offset in hex>,
+# compatible: any of ['st,stm32-spi-fifo', ...] - 'compatibe' from <binding>.yaml
+# device-type: any of ['GPIO', 'SPI', 'CAN', ...] - 'id' from <binding>.yaml
+# label: any of ['UART_0', 'SPI_11', ...] - label directive from DTS
+#
+class EDTSDatabase(Mapping):
+
+    def __init__(self, *args, **kw):
+        self._edts = dict(*args, **kw)
+        # setup basic database schema
+        for edts_key in ('devices', 'compatibles', 'device-types'):
+            if not edts_key in self._edts:
+                self._edts[edts_key] = dict()
+
+    def __getitem__(self, key):
+        return self._edts[key]
+
+    def __iter__(self):
+        return iter(self._edts)
+
+    def __len__(self):
+        return len(self._edts)
+
+    def convert_string_to_label(self, s):
+        # Transmute ,-@/ to _
+        s = s.replace("-", "_")
+        s = s.replace(",", "_")
+        s = s.replace("@", "_")
+        s = s.replace("/", "_")
+        # Uppercase the string
+        s = s.upper()
+        return s
+
+    def device_id(self, compatible, base_address, offset):
+        # sanitize
+        base_address = int(base_address)
+        offset = int(offset)
+        compatibe = self.convert_string_to_label(compatible.strip())
+        return "{}_{:08X}_{:02X}".format(compatible, base_address, offset)
+
+    def device_id_by_label(self, label):
+        for device in self._edts['devices']:
+            if label == device['label']:
+                return device['device-id']
+
+    def _update_device_compatible(self, device_id, compatible):
+        if compatible not in self._edts['compatibles']:
+            self._edts['compatibles'][compatible] = list()
+        if device_id not in self._edts['compatibles'][compatible]:
+            self._edts['compatibles'][compatible].append(device_id)
+            self._edts['compatibles'][compatible].sort()
+
+    def _update_device_type(self, device_id, device_type):
+        if device_type not in self._edts['device-types']:
+            self._edts['device-types'][device_type] = list()
+        if device_id not in self._edts['device-types'][device_type]:
+            self._edts['device-types'][device_type].append(device_id)
+            self._edts['device-types'][device_type].sort()
+
+    ##
+    # @brief Insert property value for the device of the given device id.
+    #
+    # @param device_id
```

Comment:
Using device_id will impose to generate device id each time before calling insert_device_property.
It would be easier to set node_address + compat* in the arguments and then compute device_id inside the function

*: compatible could be computed from node_address, but requires to know 'reduced'

Reply:
> Using device_id will impose to generate device id each time before calling insert_device_property.

  - On filling the EDTS database the device_id should be be generated at the beginning of extract_<whatever> and then re-used. reduced for 'compatible' is available here anyway.

```
def extract_<whatever>(node_address, ...)
    node = ....
    device_id = edts.device_id(node[props][...], ...)
``` 

There may also be a helper function in extract_dts_includes:

```
def extract_device_id(node_address):
    # extract necessary values from reduced
   return etds.device_id(compatible, base_address, offset)
```

  - On using the ETDS database the device id will be queried. Reduced is not necessary (and not available).

```
for device_id in edts.compatible_devices_id(compatible):
    # do something, eg
    label = edts.device_property('label')
```

> It would be easier to set node_address + compat* in the arguments and then compute device_id inside the function

This can not be part of EDTSDatabase as this requires reduced to be available which breaks data encapsulation. There can well be a function in extract_dts_includes that provides this, e.g.

```
def extract_edts_device_property(node_address, property_path, property_value):
    # extract necessary values from reduced
    device_id = etds.device_id(compatible, base_address, offset)
    etds.insert_device_property(device_id, property_path, property_value)
```